File: Graph-imp/cpp/cycledetectionbfs.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;


class Cycle{

     bool checkForCycle(int s,int V, vector<int> adj[], vector<int> &visited){
                
                queue<pair<int,int>> q;
                q.push({s,-1});
                visited[s] = true;

                while(!q.empty()){

                    int node = q.front().first;
                    int par =  q.front().second;
                    q.pop();

                    for(auto it: adj[node]){
                          
                          if(!visited[it]){
                              visited[it] = true;
                              q.push({it,node});
                          }

                          else if(par != it)
                             return true;

                    }
                }

        return false;
     }

  public:
      bool bfsCycleDetection(int node, vector<int> adj[]){
         vector<int> visited(node+1,0);
             
         for(int i=1; i<=node; i++){
             
             if(!visited[i]){
            if(checkForCycle(i,node, adj, visited))
                  return true;
           }
         
       }
       return false;
    }
};


int main(){
    
    int n, e;
    cin>>n>>e;
    
    vector<int> adj[n+1];
    for(int i=0; i<e; i++){
        int u,v;
        cin>>u>>v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    Cycle obj;

    bool result = obj.bfsCycleDetection(n, adj);
    
    cout<<result;
    



    return 0;
}

File: Graph-imp/java/cycledetectionbfs.java
import java.util.*;

class Cycledetectionbfs{
    public static void main(String[] args) {
    
        Scanner sc = new Scanner(System.in);
        
        int n = sc.nextInt();
        int e = sc.nextInt();
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>(n+1);
        for(int i=0; i<n+1; i++)
          adj.add(new ArrayList<Integer>());
        for(int i=0; i<e; i++){
            int u = sc.nextInt();
            int v = sc.nextInt();

            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        SolutionCycle obj = new SolutionCycle();

        boolean result = obj.bfsCycleDetection(n, adj);
        System.out.println(result);
    }
}

class Node{
    int first;
    int second;
    public Node(int first, int second){
        this.first = first;
        this.second = second;
    }
}

class SolutionCycle{

   static boolean checkCycle(int s, ArrayList<ArrayList<Integer>> adj, boolean visited[]){
           
         Queue<Node> q = new LinkedList<>();
         q.add(new Node(s,-1));

         visited[s]  = true;

         while(!q.isEmpty()){
             int node = q.peek().first;
             int par = q.peek().second;
             q.remove();

             for(Integer it : adj.get(node)){

                if(visited[it] != true){
                    visited[it] = true;
                    q.add(new Node(it,node));
                }
                else if(par != it){
                    return true;
                }
             }
         }

    return false;
   }

   public boolean bfsCycleDetection(int node, ArrayList<ArrayList<Integer>> adj){
         boolean visited[] = new boolean[node+1];

         Arrays.fill(visited, false);

         for(int i=1; i<=node; i++){

            if(visited[i] != true){

                if(checkCycle(i, adj, visited))
                     return true;
            }
         }
     return false;
   }

}